### ISSUE TITLE
`null=True, default_currency=None` on added/altered fields does not set null properly

### ISSUE THREAD

**Issue by jrocketfingers** [django-money/django-money#530]
I've added a Money field with null=True and default_currency=Non to an existing model, but it results in a SQL migration that does not make the currency field nullable. The python migration does, but it doesn't translate correctly.

```python
from django.db import migrations
import djmoney.models.fields


class Migration(migrations.Migration):

    dependencies = [
        ('core', '0001_initial'),
    ]

    operations = [
        migrations.AddField(
            model_name='examplemodelwithlateraddedmoneyfield',
            name='display',
            field=djmoney.models.fields.MoneyField(decimal_places=4, max_digits=20, null=True),
        ),
        migrations.AddField(
            model_name='examplemodelwithlateraddedmoneyfield',
            name='display_currency',
            field=djmoney.models.fields.CurrencyField(choices=[('XUA', 'ADB Unit of Account'), ('AFN', 'Afghani'), ('DZD', 'Algerian Dinar'), ('ARS', 'Argentine Peso'), ('AMD', 'Armenian Dram'), ('AWG', 'Aruban Guilder'), ('AUD', 'Australian Dollar'), ('AZN', 'Azerbaijanian Manat'), ('BSD', 'Bahamian Dollar'), ('BHD', 'Bahraini Dinar'), ('THB', 'Baht'), ('PAB', 'Balboa'), ('BBD', 'Barbados Dollar'), ('BYN', 'Belarussian Ruble'), ('BYR', 'Belarussian Ruble'), ('BZD', 'Belize Dollar'), ('BMD', 'Bermudian Dollar (customarily known as Bermuda Dollar)'), ('BTN', 'Bhutanese ngultrum'), ('VEF', 'Bolivar Fuerte'), ('BOB', 'Boliviano'), ('XBA', 'Bond Markets Units European Composite Unit (EURCO)'), ('BRL', 'Brazilian Real'), ('BND', 'Brunei Dollar'), ('BGN', 'Bulgarian Lev'), ('BIF', 'Burundi Franc'), ('XOF', 'CFA Franc BCEAO'), ('XAF', 'CFA franc BEAC'), ('XPF', 'CFP Franc'), ('CAD', 'Canadian Dollar'), ('CVE', 'Cape Verde Escudo'), ('KYD', 'Cayman Islands Dollar'), ('CLP', 'Chilean peso'), ('XTS', 'Codes specifically reserved for testing purposes'), ('COP', 'Colombian peso'), ('KMF', 'Comoro Franc'), ('CDF', 'Congolese franc'), ('BAM', 'Convertible Marks'), ('NIO', 'Cordoba Oro'), ('CRC', 'Costa Rican Colon'), ('HRK', 'Croatian Kuna'), ('CUP', 'Cuban Peso'), ('CUC', 'Cuban convertible peso'), ('CZK', 'Czech Koruna'), ('GMD', 'Dalasi'), ('DKK', 'Danish Krone'), ('MKD', 'Denar'), ('DJF', 'Djibouti Franc'), ('STD', 'Dobra'), ('DOP', 'Dominican Peso'), ('VND', 'Dong'), ('XCD', 'East Caribbean Dollar'), ('EGP', 'Egyptian Pound'), ('SVC', 'El Salvador Colon'), ('ETB', 'Ethiopian Birr'), ('EUR', 'Euro'), ('XBB', 'European Monetary Unit (E.M.U.-6)'), ('XBD', 'European Unit of Account 17(E.U.A.-17)'), ('XBC', 'European Unit of Account 9(E.U.A.-9)'), ('FKP', 'Falkland Islands Pound'), ('FJD', 'Fiji Dollar'), ('HUF', 'Forint'), ('GHS', 'Ghana Cedi'), ('GIP', 'Gibraltar Pound'), ('XAU', 'Gold'), ('XFO', 'Gold-Franc'), ('PYG', 'Guarani'), ('GNF', 'Guinea Franc'), ('GYD', 'Guyana Dollar'), ('HTG', 'Haitian gourde'), ('HKD', 'Hong Kong Dollar'), ('UAH', 'Hryvnia'), ('ISK', 'Iceland Krona'), ('INR', 'Indian Rupee'), ('IRR', 'Iranian Rial'), ('IQD', 'Iraqi Dinar'), ('IMP', 'Isle of Man Pound'), ('JMD', 'Jamaican Dollar'), ('JOD', 'Jordanian Dinar'), ('KES', 'Kenyan Shilling'), ('PGK', 'Kina'), ('LAK', 'Kip'), ('KWD', 'Kuwaiti Dinar'), ('AOA', 'Kwanza'), ('MMK', 'Kyat'), ('GEL', 'Lari'), ('LVL', 'Latvian Lats'), ('LBP', 'Lebanese Pound'), ('ALL', 'Lek'), ('HNL', 'Lempira'), ('SLL', 'Leone'), ('LSL', 'Lesotho loti'), ('LRD', 'Liberian Dollar'), ('LYD', 'Libyan Dinar'), ('SZL', 'Lilangeni'), ('LTL', 'Lithuanian Litas'), ('MGA', 'Malagasy Ariary'), ('MWK', 'Malawian Kwacha'), ('MYR', 'Malaysian Ringgit'), ('TMM', 'Manat'), ('MUR', 'Mauritius Rupee'), ('MZN', 'Metical'), ('MXV', 'Mexican Unidad de Inversion (UDI)'), ('MXN', 'Mexican peso'), ('MDL', 'Moldovan Leu'), ('MAD', 'Moroccan Dirham'), ('BOV', 'Mvdol'), ('NGN', 'Naira'), ('ERN', 'Nakfa'), ('NAD', 'Namibian Dollar'), ('NPR', 'Nepalese Rupee'), ('ANG', 'Netherlands Antillian Guilder'), ('ILS', 'New Israeli Sheqel'), ('RON', 'New Leu'), ('TWD', 'New Taiwan Dollar'), ('NZD', 'New Zealand Dollar'), ('KPW', 'North Korean Won'), ('NOK', 'Norwegian Krone'), ('PEN', 'Nuevo Sol'), ('MRO', 'Ouguiya'), ('TOP', 'Paanga'), ('PKR', 'Pakistan Rupee'), ('XPD', 'Palladium'), ('MOP', 'Pataca'), ('PHP', 'Philippine Peso'), ('XPT', 'Platinum'), ('GBP', 'Pound Sterling'), ('BWP', 'Pula'), ('QAR', 'Qatari Rial'), ('GTQ', 'Quetzal'), ('ZAR', 'Rand'), ('OMR', 'Rial Omani'), ('KHR', 'Riel'), ('MVR', 'Rufiyaa'), ('IDR', 'Rupiah'), ('RUB', 'Russian Ruble'), ('RWF', 'Rwanda Franc'), ('XDR', 'SDR'), ('SHP', 'Saint Helena Pound'), ('SAR', 'Saudi Riyal'), ('RSD', 'Serbian Dinar'), ('SCR', 'Seychelles Rupee'), ('XAG', 'Silver'), ('SGD', 'Singapore Dollar'), ('SBD', 'Solomon Islands Dollar'), ('KGS', 'Som'), ('SOS', 'Somali Shilling'), ('TJS', 'Somoni'), ('SSP', 'South Sudanese Pound'), ('LKR', 'Sri Lanka Rupee'), ('XSU', 'Sucre'), ('SDG', 'Sudanese Pound'), ('SRD', 'Surinam Dollar'), ('SEK', 'Swedish Krona'), ('CHF', 'Swiss Franc'), ('SYP', 'Syrian Pound'), ('BDT', 'Taka'), ('WST', 'Tala'), ('TZS', 'Tanzanian Shilling'), ('KZT', 'Tenge'), ('XXX', 'The codes assigned for transactions where no currency is involved'), ('TTD', 'Trinidad and Tobago Dollar'), ('MNT', 'Tugrik'), ('TND', 'Tunisian Dinar'), ('TRY', 'Turkish Lira'), ('TMT', 'Turkmenistan New Manat'), ('TVD', 'Tuvalu dollar'), ('AED', 'UAE Dirham'), ('XFU', 'UIC-Franc'), ('USD', 'US Dollar'), ('USN', 'US Dollar (Next day)'), ('UGX', 'Uganda Shilling'), ('CLF', 'Unidad de Fomento'), ('COU', 'Unidad de Valor Real'), ('UYI', 'Uruguay Peso en Unidades Indexadas (URUIURUI)'), ('UYU', 'Uruguayan peso'), ('UZS', 'Uzbekistan Sum'), ('VUV', 'Vatu'), ('CHE', 'WIR Euro'), ('CHW', 'WIR Franc'), ('KRW', 'Won'), ('YER', 'Yemeni Rial'), ('JPY', 'Yen'), ('CNY', 'Yuan Renminbi'), ('ZMK', 'Zambian Kwacha'), ('ZMW', 'Zambian Kwacha'), ('ZWD', 'Zimbabwe Dollar A/06'), ('ZWN', 'Zimbabwe dollar A/08'), ('ZWL', 'Zimbabwe dollar A/09'), ('PLN', 'Zloty')],
				default=None, editable=False, max_length=3, null=True),
        ),
    ]
```
The resulting SQL migration looks like the following:
```sql
BEGIN;                                                                             
--
-- Add field display to examplemodelwithlateraddedmoneyfield
--
ALTER TABLE "core_examplemodelwithlateraddedmoneyfield" ADD COLUMN "display" numeric(20, 4) NULL;
--
-- Add field display_currency to examplemodelwithlateraddedmoneyfield
--
ALTER TABLE "core_examplemodelwithlateraddedmoneyfield" ADD COLUMN "display_currency" varchar(3) DEFAULT 'XYZ' NOT NULL;
ALTER TABLE "core_examplemodelwithlateraddedmoneyfield" ALTER COLUMN "display_currency" DROP DEFAULT;
COMMIT;
```
The problematic part is `DEFAULT 'XYZ' NOT NULL` — I assume it should be just `NULL`? Either way, in the database the field ends up being non-nullable.

From what I've seen it should behave just as a regular CharField, so it could be related to Django version. I've tested on 2.2.7, in an example repo I'm preparing, and a few other 2.2.* versions when I first encountered it.

So far reproduced with:
Django: 2.2.*
Django Money: 0.15.1 -- 1.0
Python: 3.7.3 -- 3.8.0

Got any directions for me?

**Comment by jrocketfingers** [comment]
https://github.com/jrocketfingers/django-money-example

Here's an example project to ease reproduction.

**Comment by benjaoming** [comment]
Thanks for the thorough report!

Sounds like an issue, and maybe @Stranger6667 can recall the representation of CurrencyField and this behavior strays from the intention?

You are very welcome to propose a fix, and I can assist to review and merge changes quickly.

I recall that we observed an issue about perceived database integrity when an amount is combined with NULL currency. Something like this:

```
123 'USD': Valid
NULL NULL: Valid
NULL 'USD': Valid
123 NULL: Invalid - should not be possible
```

This was all addressed by @Stranger6667 in this PR: https://github.com/django-money/django-money/pull/261

Maybe you can use this as inspiration for a necessary fix?

**Comment by jrocketfingers** [comment]
That PR is where I found out about `default_currency=None` (should've read the docs more thoroughly 😬), and it _did_ inspire me for another future issue/proposal -- namely, DRF users don't have an ability to specify/rename currency field easily, but more on that later.

However, for this one I think I need start from SQL generating part of Django. I think we can ignore `DROP DEFAULT`, as it's explained-ish here https://code.djangoproject.com/ticket/28000, by Tim Graham (second comment).
> Django uses database defaults to set values on existing rows in a table. It doesn't leave the default values in the database so dropping the default is correct behavior. There might be an optimization not to set/drop the default in this case -- I'm not sure it's needed since the column isn't null. A separate ticket could be opened for this.

For the `NOT NULL` part, it wonks out when generating the `to_model`. I've set up a simple test case mirrored from `sqlmigrate`. [If you're still following, you can take a look here](https://github.com/jrocketfingers/django-money-example/commit/2aebbd27e3c6af2fbd8d9ad65ef32edb61097f70#diff-cf0521f6ad0d090a098a5796594e1bd4R21).

If you break in `django/db/backends/base/schema.py` in `BaseDatabaseSchemaEditor#column_sql`, you can find that `field.null` is not `True`. https://github.com/django/django/blob/stable/2.2.x/django/db/backends/base/schema.py#L157

`field` itself is passed from `django/db/migrations/operations/fields.py` in `AddField#database_forwards`.
https://github.com/django/django/blob/stable/2.2.x/django/db/migrations/operations/fields.py#L107.

This is as far as I got right now -- no idea how `to_model` is being generated.

Oh, and feel free to ignore my rambling, this is mostly so I don't forget, _and_ so someone can pick up if they encounter the same issue.

**Comment by benjaoming** [comment]
@jrocketfingers 

Are you looking at Django internals to explain the SQL migration of `CurrencyField`? Can you explain what `to_model` is doing in this analysis? Yup, I don't really understand it, but thanks for sharing and NP in elaborating your trail of investigation :+1: 

Also: Have you somehow concluded that django-money internals aren't part of this explanations? 'cause I'd think they are?

**Comment by jrocketfingers** [comment]
It's _highly likely_ that I'm missing something in django-money, but looking at `CurrencyField` I don't see anything that would affect the migration. It looks like a regular `CharField` ¯\\_(ツ)_/¯ So, am I missing something?
https://github.com/django-money/django-money/blob/master/djmoney/models/fields.py#L146

Regarding `to_model`, from what I understand it's essentially migration's view of the model after the changes are applied. It _should_ have the exact same `CurrencyField` as the field in the models.py (`null=True`), but it's `null` is set to `False`. So _something_ is being lost in translation.

However, I'm not familiar enough with django-money internals, so it could still be that. You got something specific in mind?

**Comment by benjaoming** [comment]
> However, I'm not familiar enough with django-money internals, so it could still be that. You got something specific in mind?

No, I am not getting any ideas when looking at the django-money `CurrencyField` implementation.

Could this have anything to do with `choices` being specified?

**Comment by jrocketfingers** [comment]
I didn't specify the choices, it's just the default setting -- I suppose. Anyways, I've added another field, _without_ default currency, just to be sure it doesn't affect nullability, and lo and behold, the issue still persists. I've commited the changes to the example repo.
https://github.com/jrocketfingers/django-money-example/commit/86d372ae6b0e53bf02b0708fbb4ddb9a16dc5c04#diff-e1868f46055361a7d70b4805f7b4c753R27

Output of `python django_money_example/manage.py sqlmigrate 0003`:
```
<SQL migration here>
```

So, we're back to square one, but at least we know it's just `null=True` not working correctly when a field is _added/altered_.

**Comment by mtagy** [comment]
`django-money` version 1.1 produces a totally different error.

Using `null=True, default_currency=None` generates a nullable currency field without issues in database (at least for sqlite db).
After generating migrations for the first time and applying them, every time we run the command `python manage.py makemigrations` it detects changes in model and generates a new migration file.

Something similar to the same behavior as [this 2015 resolved ticket](https://github.com/django-money/django-money/issues/106 ) but for different cause.

If I removed `default_currency=None`, no changes are detected anymore but currency field is generated as non-nullable db column with default value = `DEFAULT_CURRENCY`'s value

Environment:
django-money 1.1
django 3.0.7
python 3.7.6

**Comment by willbeaufoy** [comment]
> `django-money` version 1.1 produces a totally different error.
> 
> Using `null=True, default_currency=None` generates a nullable currency field without issues in database (at least for sqlite db).
> After generating migrations for the first time and applying them, every time we run the command `python manage.py makemigrations` it detects changes in model and generates a new migration file.
> 
> Something similar to the same behavior as [this 2015 resolved ticket](https://github.com/django-money/django-money/issues/106) but for different cause.
> 
> If I removed `default_currency=None`, no changes are detected anymore but currency field is generated as non-nullable db column with default value = `DEFAULT_CURRENCY`'s value.
> 
> Environment:
> django-money 1.1
> django 3.0.7
> python 3.7.6
> 
> _Note_, I didn't know if I should open a new issue for this or not, that's why I left a comment for now.
> If required, I can report it as a new issue.

Was a new issue for this ever opened? It is still a problem for me.

**Comment by jrocketfingers** [comment]
I never got around to fixing it :( I hope my notes are detailed enough to get back into it and fix it, but I can't promise anything. Wish there was a way for open source work to be done in pairs, cause I lost it last time I tried tackling this.

**Comment by deejax** [comment]
I am facing both the issues :( . The one mentioned by @willbeaufoy  and the original issue reported by @jrocketfingers .

**Comment by flaeppe** [comment]
A somewhat of a workaround when one is interested in _raising an exception_ when a row is inserted with the default value, though not solved via the field definition and any `default=None`, is to utilise [CheckConstraint](https://docs.djangoproject.com/en/dev/ref/models/constraints/#checkconstraint) when on `Django >= 2.2`. And avoid touching the `default_currency=`, set it to `default_currency="XYZ"` or something else that'll violate the constraint.

Example:

```python
# models.py
class MoneyModel(...):
    money = MoneyField(
        max_digits=5,
        decimal_places=2,
        # Specifying currency_choices allows for error messages in admin, etc.
        currency_choices=[("EUR", "EUR"), ("USD", "USD")],
    )

    class Meta:
        constraints = [
            CheckConstraint(check=Q(money_currency__in=["EUR", "USD"]), name="currency_check")
        ]
```

And since `MoneyField` will set: `default="XYZ"` on the generated currency field the migration will do:

```python
# migration.py
class Migration(migrations.Migration):
    ...
    operations = [
        migrations.CreateModel(
            name="MoneyModel",
            fields=[
                ...
                ("money_currency", djmoney.models.fields.CurrencyField(..., default="XYZ", ...)),
                ...
            ]
        ),
        migrations.AddConstraint(
            model_name="moneymodel",
            constraint=models.CheckConstraint(
                check=models.Q(("money_currency__in", ["EUR", "USD"])),
                name="currency_check",
            ),
        ),

    ]
```

**Comment by flaeppe** [comment]
> > `django-money` version 1.1 produces a totally different error.
> > Using `null=True, default_currency=None` generates a nullable currency field without issues in database (at least for sqlite db).
> > After generating migrations for the first time and applying them, every time we run the command `python manage.py makemigrations` it detects changes in model and generates a new migration file.
> > Something similar to the same behavior as [this 2015 resolved ticket](https://github.com/django-money/django-money/issues/106) but for different cause.
> > If I removed `default_currency=None`, no changes are detected anymore but currency field is generated as non-nullable db column with default value = `DEFAULT_CURRENCY`'s value.
> > Environment:
> > django-money 1.1
> > django 3.0.7
> > python 3.7.6
> > _Note_, I didn't know if I should open a new issue for this or not, that's why I left a comment for now.
> > If required, I can report it as a new issue.
> 
> Was a new issue for this ever opened? It is still a problem for me on 1.3.1, and I don't see any commits in 2.0 relating to it.

I've opened #646 together with a fix in #647

**Comment by flaeppe** [comment]
I've tried to reproduce the invalid SQL seen in https://github.com/django-money/django-money/issues/530#issue-527688265 with the changes from #647 applied. But I haven't been able to do so. Something tells me the seen SQL `DEFAULT` value comes from `DEFAULT_CURRENCY` and I'm fairly certain that the invalid SQL was a direct cause of the invalid deconstruction that was resolved in #647.

```python
# migration.py
from django.db import migrations
import djmoney.models.fields


class Migration(migrations.Migration):

    dependencies = [
        ('money_app', '0001_initial'),
    ]

    operations = [
        migrations.AddField(
            model_name='moneytable',
            name='second_money',
            field=djmoney.models.fields.MoneyField(decimal_places=4, default_currency=None, max_digits=20, null=True),
        ),
        migrations.AddField(
            model_name='moneytable',
            name='second_money_currency',
            field=djmoney.models.fields.CurrencyField(
                choices=[('XUA', 'ADB Unit of Account'), ('AFN', 'Afghan Afghani'), ('AFA', 'Afghan Afghani (1927–2002)'), ('ALL', 'Albanian Lek'), ('ALK', 'Albanian Lek (1946–1965)'), ('DZD', 'Algerian Dinar'), ('ADP', 'Andorran Peseta'), ('AOA', 'Angolan Kwanza'), ('AOK', 'Angolan Kwanza (1977–1991)'), ('AON', 'Angolan New Kwanza (1990–2000)'), ('AOR', 'Angolan Readjusted Kwanza (1995–1999)'), ('ARA', 'Argentine Austral'), ('ARS', 'Argentine Peso'), ('ARM', 'Argentine Peso (1881–1970)'), ('ARP', 'Argentine Peso (1983–1985)'), ('ARL', 'Argentine Peso Ley (1970–1983)'), ('AMD', 'Armenian Dram'), ('AWG', 'Aruban Florin'), ('AUD', 'Australian Dollar'), ('ATS', 'Austrian Schilling'), ('AZN', 'Azerbaijani Manat'), ('AZM', 'Azerbaijani Manat (1993–2006)'), ('BSD', 'Bahamian Dollar'), ('BHD', 'Bahraini Dinar'), ('BDT', 'Bangladeshi Taka'), ('BBD', 'Barbadian Dollar'), ('BYN', 'Belarusian Ruble'), ('BYB', 'Belarusian Ruble (1994–1999)'), ('BYR', 'Belarusian Ruble (2000–2016)'), ('BEF', 'Belgian Franc'), ('BEC', 'Belgian Franc (convertible)'), ('BEL', 'Belgian Franc (financial)'), ('BZD', 'Belize Dollar'), ('BMD', 'Bermudan Dollar'), ('BTN', 'Bhutanese Ngultrum'), ('BOB', 'Bolivian Boliviano'), ('BOL', 'Bolivian Boliviano (1863–1963)'), ('BOV', 'Bolivian Mvdol'), ('BOP', 'Bolivian Peso'), ('BAM', 'Bosnia-Herzegovina Convertible Mark'), ('BAD', 'Bosnia-Herzegovina Dinar (1992–1994)'), ('BAN', 'Bosnia-Herzegovina New Dinar (1994–1997)'), ('BWP', 'Botswanan Pula'), ('BRC', 'Brazilian Cruzado (1986–1989)'), ('BRZ', 'Brazilian Cruzeiro (1942–1967)'), ('BRE', 'Brazilian Cruzeiro (1990–1993)'), ('BRR', 'Brazilian Cruzeiro (1993–1994)'), ('BRN', 'Brazilian New Cruzado (1989–1990)'), ('BRB', 'Brazilian New Cruzeiro (1967–1986)'), ('BRL', 'Brazilian Real'), ('GBP', 'British Pound'), ('BND', 'Brunei Dollar'), ('BGL', 'Bulgarian Hard Lev'), ('BGN', 'Bulgarian Lev'), ('BGO', 'Bulgarian Lev (1879–1952)'), ('BGM', 'Bulgarian Socialist Lev'), ('BUK', 'Burmese Kyat'), ('BIF', 'Burundian Franc'), ('XPF', 'CFP Franc'), ('KHR', 'Cambodian Riel'), ('CAD', 'Canadian Dollar'), ('CVE', 'Cape Verdean Escudo'), ('KYD', 'Cayman Islands Dollar'), ('XAF', 'Central African CFA Franc'), ('CLE', 'Chilean Escudo'), ('CLP', 'Chilean Peso'), ('CLF', 'Chilean Unit of Account (UF)'), ('CNX', 'Chinese People’s Bank Dollar'), ('CNY', 'Chinese Yuan'), ('CNH', 'Chinese Yuan (offshore)'), ('COP', 'Colombian Peso'), ('COU', 'Colombian Real Value Unit'), ('KMF', 'Comorian Franc'), ('CDF', 'Congolese Franc'), ('CRC', 'Costa Rican Colón'), ('HRD', 'Croatian Dinar'), ('HRK', 'Croatian Kuna'), ('CUC', 'Cuban Convertible Peso'), ('CUP', 'Cuban Peso'), ('CYP', 'Cypriot Pound'), ('CZK', 'Czech Koruna'), ('CSK', 'Czechoslovak Hard Koruna'), ('DKK', 'Danish Krone'), ('DJF', 'Djiboutian Franc'), ('DOP', 'Dominican Peso'), ('NLG', 'Dutch Guilder'), ('XCD', 'East Caribbean Dollar'), ('DDM', 'East German Mark'), ('ECS', 'Ecuadorian Sucre'), ('ECV', 'Ecuadorian Unit of Constant Value'), ('EGP', 'Egyptian Pound'), ('GQE', 'Equatorial Guinean Ekwele'), ('ERN', 'Eritrean Nakfa'), ('EEK', 'Estonian Kroon'), ('ETB', 'Ethiopian Birr'), ('EUR', 'Euro'), ('XBA', 'European Composite Unit'), ('XEU', 'European Currency Unit'), ('XBB', 'European Monetary Unit'), ('XBC', 'European Unit of Account (XBC)'), ('XBD', 'European Unit of Account (XBD)'), ('FKP', 'Falkland Islands Pound'), ('FJD', 'Fijian Dollar'), ('FIM', 'Finnish Markka'), ('FRF', 'French Franc'), ('XFO', 'French Gold Franc'), ('XFU', 'French UIC-Franc'), ('GMD', 'Gambian Dalasi'), ('GEK', 'Georgian Kupon Larit'), ('GEL', 'Georgian Lari'), ('DEM', 'German Mark'), ('GHS', 'Ghanaian Cedi'), ('GHC', 'Ghanaian Cedi (1979–2007)'), ('GIP', 'Gibraltar Pound'), ('XAU', 'Gold'), ('GRD', 'Greek Drachma'), ('GTQ', 'Guatemalan Quetzal'), ('GWP', 'Guinea-Bissau Peso'), ('GNF', 'Guinean Franc'), ('GNS', 'Guinean Syli'), ('GYD', 'Guyanaese Dollar'), ('HTG', 'Haitian Gourde'), ('HNL', 'Honduran Lempira'), ('HKD', 'Hong Kong Dollar'), ('HUF', 'Hungarian Forint'), ('IMP', 'IMP'), ('ISK', 'Icelandic Króna'), ('ISJ', 'Icelandic Króna (1918–1981)'), ('INR', 'Indian Rupee'), ('IDR', 'Indonesian Rupiah'), ('IRR', 'Iranian Rial'), ('IQD', 'Iraqi Dinar'), ('IEP', 'Irish Pound'), ('ILS', 'Israeli New Shekel'), ('ILP', 'Israeli Pound'), ('ILR', 'Israeli Shekel (1980–1985)'), ('ITL', 'Italian Lira'), ('JMD', 'Jamaican Dollar'), ('JPY', 'Japanese Yen'), ('JOD', 'Jordanian Dinar'), ('KZT', 'Kazakhstani Tenge'), ('KES', 'Kenyan Shilling'), ('KWD', 'Kuwaiti Dinar'), ('KGS', 'Kyrgystani Som'), ('LAK', 'Laotian Kip'), ('LVL', 'Latvian Lats'), ('LVR', 'Latvian Ruble'), ('LBP', 'Lebanese Pound'), ('LSL', 'Lesotho Loti'), ('LRD', 'Liberian Dollar'), ('LYD', 'Libyan Dinar'), ('LTL', 'Lithuanian Litas'), ('LTT', 'Lithuanian Talonas'), ('LUL', 'Luxembourg Financial Franc'), ('LUC', 'Luxembourgian Convertible Franc'), ('LUF', 'Luxembourgian Franc'), ('MOP', 'Macanese Pataca'), ('MKD', 'Macedonian Denar'), ('MKN', 'Macedonian Denar (1992–1993)'), ('MGA', 'Malagasy Ariary'), ('MGF', 'Malagasy Franc'), ('MWK', 'Malawian Kwacha'), ('MYR', 'Malaysian Ringgit'), ('MVR', 'Maldivian Rufiyaa'), ('MVP', 'Maldivian Rupee (1947–1981)'), ('MLF', 'Malian Franc'), ('MTL', 'Maltese Lira'), ('MTP', 'Maltese Pound'), ('MRU', 'Mauritanian Ouguiya'), ('MRO', 'Mauritanian Ouguiya (1973–2017)'), ('MUR', 'Mauritian Rupee'), ('MXV', 'Mexican Investment Unit'), ('MXN', 'Mexican Peso'), ('MXP', 'Mexican Silver Peso (1861–1992)'), ('MDC', 'Moldovan Cupon'), ('MDL', 'Moldovan Leu'), ('MCF', 'Monegasque Franc'), ('MNT', 'Mongolian Tugrik'), ('MAD', 'Moroccan Dirham'), ('MAF', 'Moroccan Franc'), ('MZE', 'Mozambican Escudo'), ('MZN', 'Mozambican Metical'), ('MZM', 'Mozambican Metical (1980–2006)'), ('MMK', 'Myanmar Kyat'), ('NAD', 'Namibian Dollar'), ('NPR', 'Nepalese Rupee'), ('ANG', 'Netherlands Antillean Guilder'), ('TWD', 'New Taiwan Dollar'), ('NZD', 'New Zealand Dollar'), ('NIO', 'Nicaraguan Córdoba'), ('NIC', 'Nicaraguan Córdoba (1988–1991)'), ('NGN', 'Nigerian Naira'), ('KPW', 'North Korean Won'), ('NOK', 'Norwegian Krone'), ('OMR', 'Omani Rial'), ('PKR', 'Pakistani Rupee'), ('XPD', 'Palladium'), ('PAB', 'Panamanian Balboa'), ('PGK', 'Papua New Guinean Kina'), ('PYG', 'Paraguayan Guarani'), ('PEI', 'Peruvian Inti'), ('PEN', 'Peruvian Sol'), ('PES', 'Peruvian Sol (1863–1965)'), ('PHP', 'Philippine Piso'), ('XPT', 'Platinum'), ('PLN', 'Polish Zloty'), ('PLZ', 'Polish Zloty (1950–1995)'), ('PTE', 'Portuguese Escudo'), ('GWE', 'Portuguese Guinea Escudo'), ('QAR', 'Qatari Rial'), ('XRE', 'RINET Funds'), ('RHD', 'Rhodesian Dollar'), ('RON', 'Romanian Leu'), ('ROL', 'Romanian Leu (1952–2006)'), ('RUB', 'Russian Ruble'), ('RUR', 'Russian Ruble (1991–1998)'), ('RWF', 'Rwandan Franc'), ('SVC', 'Salvadoran Colón'), ('WST', 'Samoan Tala'), ('SAR', 'Saudi Riyal'), ('RSD', 'Serbian Dinar'), ('CSD', 'Serbian Dinar (2002–2006)'), ('SCR', 'Seychellois Rupee'), ('SLL', 'Sierra Leonean Leone'), ('XAG', 'Silver'), ('SGD', 'Singapore Dollar'), ('SKK', 'Slovak Koruna'), ('SIT', 'Slovenian Tolar'), ('SBD', 'Solomon Islands Dollar'), ('SOS', 'Somali Shilling'), ('ZAR', 'South African Rand'), ('ZAL', 'South African Rand (financial)'), ('KRH', 'South Korean Hwan (1953–1962)'), ('KRW', 'South Korean Won'), ('KRO', 'South Korean Won (1945–1953)'), ('SSP', 'South Sudanese Pound'), ('SUR', 'Soviet Rouble'), ('ESP', 'Spanish Peseta'), ('ESA', 'Spanish Peseta (A account)'), ('ESB', 'Spanish Peseta (convertible account)'), ('XDR', 'Special Drawing Rights'), ('LKR', 'Sri Lankan Rupee'), ('SHP', 'St. Helena Pound'), ('XSU', 'Sucre'), ('SDD', 'Sudanese Dinar (1992–2007)'), ('SDG', 'Sudanese Pound'), ('SDP', 'Sudanese Pound (1957–1998)'), ('SRD', 'Surinamese Dollar'), ('SRG', 'Surinamese Guilder'), ('SZL', 'Swazi Lilangeni'), ('SEK', 'Swedish Krona'), ('CHF', 'Swiss Franc'), ('SYP', 'Syrian Pound'), ('STN', 'São Tomé & Príncipe Dobra'), ('STD', 'São Tomé & Príncipe Dobra (1977–2017)'), ('TVD', 'TVD'), ('TJR', 'Tajikistani Ruble'), ('TJS', 'Tajikistani Somoni'), ('TZS', 'Tanzanian Shilling'), ('XTS', 'Testing Currency Code'), ('THB', 'Thai Baht'), ('XXX', 'The codes assigned for transactions where no currency is involved'), ('TPE', 'Timorese Escudo'), ('TOP', 'Tongan Paʻanga'), ('TTD', 'Trinidad & Tobago Dollar'), ('TND', 'Tunisian Dinar'), ('TRY', 'Turkish Lira'), ('TRL', 'Turkish Lira (1922–2005)'), ('TMT', 'Turkmenistani Manat'), ('TMM', 'Turkmenistani Manat (1993–2009)'), ('USD', 'US Dollar'), ('USN', 'US Dollar (Next day)'), ('USS', 'US Dollar (Same day)'), ('UGX', 'Ugandan Shilling'), ('UGS', 'Ugandan Shilling (1966–1987)'), ('UAH', 'Ukrainian Hryvnia'), ('UAK', 'Ukrainian Karbovanets'), ('AED', 'United Arab Emirates Dirham'), ('UYW', 'Uruguayan Nominal Wage Index Unit'), ('UYU', 'Uruguayan Peso'), ('UYP', 'Uruguayan Peso (1975–1993)'), ('UYI', 'Uruguayan Peso (Indexed Units)'), ('UZS', 'Uzbekistani Som'), ('VUV', 'Vanuatu Vatu'), ('VES', 'Venezuelan Bolívar'), ('VEB', 'Venezuelan Bolívar (1871–2008)'), ('VEF', 'Venezuelan Bolívar (2008–2018)'), ('VND', 'Vietnamese Dong'), ('VNN', 'Vietnamese Dong (1978–1985)'), ('CHE', 'WIR Euro'), ('CHW', 'WIR Franc'), ('XOF', 'West African CFA Franc'), ('YDD', 'Yemeni Dinar'), ('YER', 'Yemeni Rial'), ('YUN', 'Yugoslavian Convertible Dinar (1990–1992)'), ('YUD', 'Yugoslavian Hard Dinar (1966–1990)'), ('YUM', 'Yugoslavian New Dinar (1994–2002)'), ('YUR', 'Yugoslavian Reformed Dinar (1992–1993)'), ('ZWN', 'ZWN'), ('ZRN', 'Zairean New Zaire (1993–1998)'), ('ZRZ', 'Zairean Zaire (1971–1993)'), ('ZMW', 'Zambian Kwacha'), ('ZMK', 'Zambian Kwacha (1968–2012)'), ('ZWD', 'Zimbabwean Dollar (1980–2008)'), ('ZWR', 'Zimbabwean Dollar (2008)'), ('ZWL', 'Zimbabwean Dollar (2009)')],
                default=None, editable=False, max_length=3, null=True
            ),
        ),
    ]
```

Here's the generated SQL from the above migration

```sql
BEGIN;
--
-- Add field second_money to moneytable
--
ALTER TABLE "money_app_moneytable" ADD COLUMN "second_money" numeric(20, 4) NULL;
--
-- Add field second_money_currency to moneytable
--
ALTER TABLE "money_app_moneytable" ADD COLUMN "second_money_currency" varchar(3) NULL;
COMMIT;
```

For reference, this is the model declaration

```python
# models.py
from django.db import models
from djmoney.money import Money
from djmoney.models.fields import MoneyField


class MoneyTable(models.Model):
    money = MoneyField(max_digits=9, null=True, default_currency=None)
    # Uncomment the field below after initial migration is generated, so that it's
    # created via an AddField migration operation
    # second_money = MoneyField(max_digits=20, decimal_places=4, null=True, default_currency=None)
```

---

Any input on if I've missed something is appreciated.

**Comment by jrocketfingers** [comment]
I just realized I've never closed this as fixed, so here we go. Thanks everyone!
